### PR TITLE
[Parser] Completely rewrite the parser, using a combinator approach 

### PR DESCRIPTION
**Pull request by Hazurl**
- Using Parser combinators, a way of constructing bigger parsers from smaller one
- [Documentation about the parsers](https://github.com/WotScript/parser/blob/356ef28eaf02cd951cdae38f8112041fc18d670e/include/ws/parser/ParserInternal.hpp#L16)
- Printable AST
- Cleaner project using folders
- Create macro for Test (unreadable now ? but shorter at least; the best would be to have our own lexer...)

Thing to be done:
- Basic lexer for the tests (just replace characters by a token)
- Errors, ParserError need to change, it also need to be able to combine two errors
- Remove all magic strings, translation from (sub)type to string need a function and also the conversion from (sub)type to string

### CODE REVIEW

File: src/test.cpp [hunk 1]
```diff
@@ -26,6 +26,23 @@ ws::parser::Token sub() {
...
+
+#define CHECK(parsable, print_ast, is...) check({ is }, parsable, print_ast)
+#define P plus()
+#define S sub()
+#define M mult()
+#define D div()
+#define L left()
+#define R right()
+#define N(n) number(n)
+
```
[the remark this answers is not in the document]
Comment (Repertoi-e) [reply src/test.cpp]: I think they are OK, they are used only in `test.cpp`, not global.

File: src/test.cpp [hunk 2]
```diff
@@ -46,7 +63,7 @@ bool check(std::vector<ws::parser::Token> const& tokens, bool parsable, bool pri
     bool test_pass = !is_error(out) == parsable;
 
     if (test_pass)
-        ws::module::print(ws::module::colour::fg::green, ws::module::style::bold, "[O]", ws::module::style::reset, " OK");
+        ws::module::print(ws::module::colour::fg::green, ws::module::style::bold, "[_]", ws::module::style::reset, " OK");
```
[the remark this answers is not in the document]
Comment (Hazurl) [reply src/test.cpp]: gonna update the lib later

File: src/test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hazurl) [reply src/test.cpp]: I want to make a little lexer so it looks like `tokenize("a+a--a/a")`

**Comment by Hazurl** [comment]
@Repertoi-e I wait for your review ?

File: src/test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jackojc) [reply src/test.cpp]: https://www.youtube.com/watch?v=lD_ag67tH3I

**Pull request merged**